### PR TITLE
Support passing FileOpener for local cache

### PR DESCRIPTION
**Pull request by apc999**

### CODE REVIEW

File: core/client/hdfs/src/main/java/alluxio/hadoop/LocalCacheFileSystem.java [hunk 1]
```diff
@@ -112,10 +121,8 @@ public FSDataInputStream open(Path path, int bufferSize) throws IOException {
     if (mCacheManager == null) {
       return mExternalFileSystem.open(path, bufferSize);
     }
-    return new FSDataInputStream(new HdfsFileInputStream(
-        new LocalCacheFileInStream(new AlluxioURI(path.toString()),
-            OpenFilePOptions.getDefaultInstance(), mExternalFileSystemWrapper, mCacheManager),
-        statistics));
+    URIStatus status = HadoopUtils.toAlluxioUriStatus(mExternalFileSystem.getFileStatus(path));
+    return open(status, bufferSize);
```
Comment (calvinjia) [inline core/client/hdfs/src/main/java/alluxio/hadoop/LocalCacheFileSystem.java:125]: Do we only need 2 of these three methods? ie. the interface `open(Path, int)` and implementation `open(URIStatus, int)`? We should always use the hadoop file opener if cache is not enabled?

Also, buffer size is only respected if the cache is not enabled?

**Review by calvinjia — CHANGES_REQUESTED** [review]

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileSystem.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public FileInStream openFile(AlluxioURI path, OpenFilePOptions options)
     if (mCacheManager == null || mCacheManager.state() == CacheManager.State.NOT_IN_USE) {
       return mDelegatedFileSystem.openFile(path, options);
     }
-    return new LocalCacheFileInStream(path, options, mDelegatedFileSystem, mCacheManager);
+    return openFile(mDelegatedFileSystem.getStatus(path), options);
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileSystem.java:59]: it was not called lazily even prior to this pr.
We memorized a closure and call `get` right away. 
it actually caused issues in error handling: exceptions are wrapped as RTE to through which is not nice.

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 3]
```diff
@@ -48,13 +46,11 @@
...
   /** Local store to store pages. */
   private final CacheManager mCacheManager;
-  /** External storage system. */
-  private final FileSystem mExternalFs;
   /** Path of the file. */
   private final AlluxioURI mPath;
   /** File info, fetched from external FS. */
   private final URIStatus mStatus;
-  private final OpenFilePOptions mOpenOptions;
+  private final FileInStreamOpener mFileInStreamOpener;
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:53]: done

File: core/client/hdfs/src/main/java/alluxio/hadoop/LocalCacheFileSystem.java (hunk 1 above)
Comment (apc999) [reply core/client/hdfs/src/main/java/alluxio/hadoop/LocalCacheFileSystem.java:125]: > Do we only need 2 of these three methods? ie. the interface open(Path, int) and implementation open(URIStatus, int)? We should always use the hadoop file opener if cache is not enabled?

correct, i removed the 3rd `open` method with `FileOpener`. originally I thought I needed in Presto land, but it turns out not the case (see draft PR https://github.com/prestodb/presto/pull/15344/files)

> Also, buffer size is only respected if the cache is not enabled?

Correct

**Comment by apc999** [comment]
@calvinjia  PTAL

**Review by calvinjia — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by apc999** [comment]
alluxio-bot, cherry-pick this to branch-2.4 please

**Pull request merged**